processing/src/main/java/org/apache/druid/segment/AutoTypeColumnIndexer.java

```diff
@@ -133,6 +150,51 @@ public EncodedKeyComponent<StructuredData> processRowValsToUnsortedEncodedKeyCom
     } else if (isConstant) {
       isConstant = Objects.equals(dimValues, constantValue);
     }
+
+    if (castToExpressionType != null) {
+      return processCast(dimValues);
+    } else {
+      return processAuto(dimValues);
+    }
+  }
+
+  /**
+   * Process values which will all be cast to {@link #castToExpressionType}. This method should not be used for
+   * and does not handle actual nested data structures, use {@link #processAuto(Object)} instead.
+   */
+  private EncodedKeyComponent<StructuredData> processCast(@Nullable Object dimValues)
+  {
+    final long oldDictSizeInBytes = globalDictionary.sizeInBytes();
+    final int oldFieldKeySize = estimatedFieldKeySize;
+    ExprEval<?> eval = ExprEval.bestEffortOf(dimValues);
+    try {
+      eval = eval.castTo(castToExpressionType);
+      FieldIndexer fieldIndexer = fieldIndexers.get(NestedPathFinder.JSON_PATH_ROOT);
+      if (fieldIndexer == null) {
+        estimatedFieldKeySize += StructuredDataProcessor.estimateStringSize(NestedPathFinder.JSON_PATH_ROOT);
+        fieldIndexer = new FieldIndexer(globalDictionary);
+        fieldIndexers.put(NestedPathFinder.JSON_PATH_ROOT, fieldIndexer);
+      }
+      StructuredDataProcessor.ProcessedValue<?> rootValue = fieldIndexer.processValue(eval);
+      long effectiveSizeBytes = rootValue.getSize();
+      // then, we add the delta of size change to the global dictionaries to account for any new space added by the
+      // 'raw' data
+      effectiveSizeBytes += (globalDictionary.sizeInBytes() - oldDictSizeInBytes);
+      effectiveSizeBytes += (estimatedFieldKeySize - oldFieldKeySize);
+      return new EncodedKeyComponent<>(StructuredData.wrap(eval.value()), effectiveSizeBytes);
+    }
+    catch (IAE invalidCast) {
+      throw new ParseException(eval.asString(), invalidCast, "Cannot coerce to requested type [%s]", castToType);
```

Comment:
can we include the field / column name here?

Reply:
seems reasonable, but will make sure whatever catches these doesn't already do something like that